### PR TITLE
dhcp: introduce sd_dhcp_client_id and relevant functions

### PR DESCRIPTION
**Pull request by yuwata**
split-out of #30021.

### CODE REVIEW

File: src/libsystemd-network/sd-dhcp-client-id.c [hunk 1]
```diff
@@ -0,0 +1,95 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include "alloc-util.h"
+#include "dhcp-client-id-internal.h"
+#include "unaligned.h"
+#include "utf8.h"
+
+int sd_dhcp_client_id_clear(sd_dhcp_client_id *client_id) {
```
Comment (poettering) [inline src/libsystemd-network/sd-dhcp-client-id.c:8]: we usually call this _done() or _reset(), no? (didn't check, we might be a bit chaotic there)

File: src/libsystemd-network/sd-dhcp-client-id.c [hunk 2]
```diff
@@ -93,3 +93,67 @@ int sd_dhcp_client_id_set_iaid_duid(
...
+        size_t len;
+        int r;
+
+        assert_return(sd_dhcp_client_id_is_set(client_id), -EINVAL);
+        assert_return(ret, -EINVAL);
+
+        len = client_id->size - offsetof(typeof(client_id->id), data);
+
+        switch (client_id->id.type) {
+        case 0:
```
Comment (poettering) [inline src/libsystemd-network/sd-dhcp-client-id.c:108]: no named enum values for the cases of this switch?

File: src/libsystemd-network/sd-dhcp-lease.c [hunk 3]
```diff
@@ -1066,8 +1065,8 @@ int dhcp_lease_save(sd_dhcp_lease *lease, const char *lease_file) {
         _cleanup_fclose_ FILE *f = NULL;
         struct in_addr address;
         const struct in_addr *addresses;
-        const void *client_id, *data;
-        size_t client_id_len, data_len;
+        const void *data;
+        size_t data_len;
```
Comment (poettering) [inline src/libsystemd-network/sd-dhcp-lease.c:1069]: btw, in the tpm2 code i started to switch cases like this (where you have some arbitrary blob pointer plus size) to use "struct iovec". There are a bunch of helpers to make working with such pointer+size pairs easy in iovec-util.h.

Might be somewthing to use here too, but doesn't matter

File: src/libsystemd-network/sd-dhcp-client-id.c (hunk 1 above)
Comment (yuwata) [reply src/libsystemd-network/sd-dhcp-client-id.c:8]: `_done()` is usually (not always) used when the object will not be reused.
`_clear()` makes the passed object reused.
But, such rule is not uniformly applied in existing code, IIRC.
Personally, I prefer `_clear()` here.

File: src/libsystemd-network/sd-dhcp-client-id.c (hunk 2 above)
Comment (yuwata) [reply src/libsystemd-network/sd-dhcp-client-id.c:108]: I will add an enum for the type in a later PR.

File: src/libsystemd-network/sd-dhcp-lease.c (hunk 3 above)
Comment (yuwata) [reply src/libsystemd-network/sd-dhcp-lease.c:1069]: Sounds nice. But I'd like to do that consistently. I will try that in a later PR.

**Pull request merged**